### PR TITLE
Add CopyObject Function #128

### PR DESCRIPTION
**Pull request by robertmin1**
Add CopyObject Function #128

### CODE REVIEW

**Comment by JeremyRand** [comment]
Well, I'm disappointed that some Firefox tests are still failing when I thought I fixed that, but that doesn't look like it's this PR's fault.

**Comment by JeremyRand** [comment]
The "Go Lint New" failure affects the whole repo, and I won't make you fix that in this PR.

File: p11mod/p11mod.go [hunk 1]
```diff
@@ -410,9 +410,29 @@ func (ll *llBackend) CreateObject(sh pkcs11.SessionHandle, temp []*pkcs11.Attrib
...
-	return 0, pkcs11.Error(pkcs11.CKR_FUNCTION_NOT_SUPPORTED)
+	session, err := ll.getSessionByHandle(sh)
+	if err != nil {
+		return 0, err
+	}
+
+	objectIndex := int(o - 1)
+
+	if objectIndex < 0 || objectIndex >= len(session.objects) {
+		log.Printf("p11mod CopyObject: index invalid: requested %d, object count %d\n", objectIndex, len(session.objects))
```
Comment (JeremyRand) [inline p11mod/p11mod.go:421]: Please return an error in this case; logging is good but not sufficient.

File: p11mod/p11mod.go [hunk 2]
```diff
@@ -410,9 +410,29 @@ func (ll *llBackend) CreateObject(sh pkcs11.SessionHandle, temp []*pkcs11.Attrib
...
 func (ll *llBackend) CopyObject(sh pkcs11.SessionHandle, o pkcs11.ObjectHandle, temp []*pkcs11.Attribute) (pkcs11.ObjectHandle, error) {
-	// TODO
-	log.Println("p11mod CopyObject: not implemented")
-	return 0, pkcs11.Error(pkcs11.CKR_FUNCTION_NOT_SUPPORTED)
+	session, err := ll.getSessionByHandle(sh)
+	if err != nil {
+		return 0, err
+	}
+
+	objectIndex := int(o - 1)
```
Comment (JeremyRand) [inline p11mod/p11mod.go:418]: Please add a comment explaining this line, similar to what `DestroyObject` has.

File: p11mod/p11mod.go [hunk 3]
```diff
@@ -410,9 +410,29 @@ func (ll *llBackend) CreateObject(sh pkcs11.SessionHandle, temp []*pkcs11.Attrib
...
+	object := session.objects[objectIndex]
+
+	objectc, err := object.Copy(temp)
+	if err != nil {
+		return 0, err
+	}
+
+	session.objects = append(session.objects, objectc)
+
+	resultHandle := len(session.objects)
```
Comment (JeremyRand) [inline p11mod/p11mod.go:433]: Please add a comment explaining this line, similar to what CreateObject has.

**Comment by JeremyRand** [comment]
ACK 48a757a4c09bdd4ce0d0437fd881ec4af7675ab6 , modulo the 3 comments above.

File: p11mod/p11mod.go (hunk 2 above)
Comment (robertmin1) [reply p11mod/p11mod.go:419]: Comment Added

File: p11mod/p11mod.go (hunk 3 above)
Comment (robertmin1) [reply p11mod/p11mod.go:434]: Comment Added

File: p11mod/p11mod.go (hunk 1 above)
Comment (robertmin1) [reply p11mod/p11mod.go:422]: return error added

**Comment by robertmin1** [comment]
I'll check on Go Lint New, to determine what is causing the failures

**Comment by robertmin1** [comment]
Go Lint New returning an error: "return with no blank line before" But the return statement is alone inside a statement group.
Weird.

File: p11mod/p11mod.go [hunk 4]
```diff
@@ -410,9 +411,34 @@ func (ll *llBackend) CreateObject(sh pkcs11.SessionHandle, temp []*pkcs11.Attrib
...
+	if err != nil {
+		return 0, err
+	}
+
+	// Handles are 1-indexed, while our slice is 0-indexed.
+	objectIndex := int(o - 1)
+
+	if objectIndex < 0 || objectIndex >= len(session.objects) {
+		log.Printf("p11mod CopyObject: index invalid: requested %d, object count %d\n", objectIndex, len(session.objects))
+		return 0, pkcs11.Error(pkcs11.CKR_FUNCTION_NOT_SUPPORTED)
```
Comment (JeremyRand) [inline p11mod/p11mod.go:424]: I think this is the wrong error code?

Comment (JeremyRand) [inline p11mod/p11mod.go:424]: Putting a blank line between the above `log.Printf` and this line should fix the linter error.

File: .gitignore [hunk 5]
```diff
@@ -3,3 +3,5 @@
 /*.a
 /*.o
 /moz/web-ext-artifacts/
+go.mod
+go.sum
```
Comment (JeremyRand) [inline .gitignore:7]: Was it intentional that a `.gitignore` change got added to this PR?

Comment (robertmin1) [reply .gitignore:7]: It is from the merge it did.

Comment (robertmin1) [reply .gitignore]: Fixed, updated the branch!

File: p11mod/p11mod.go (hunk 4 above)
Comment (robertmin1) [reply p11mod/p11mod.go]: This doesn't fix the error, the test passed then failed on the second try.

Comment (robertmin1) [reply p11mod/p11mod.go]: Let me Re-run it

Comment (robertmin1) [reply p11mod/p11mod.go]: Go Lint new test passed

**Comment by robertmin1** [comment]
Accidentally closed the PR when force pushing :(

**Pull request closed without merging**